### PR TITLE
Process reviewer invitation on login

### LINKED ISSUE #710: process invited reviewer on first sign in
part of #253 

- [ ] saml
- [ ] registered

### PR DESCRIPTION
**Pull request by jseraf**
Check for reviewer_invitations on user create. 

By-pass devise :confirmable for invited registered_user

part of #253 
closes #710 
related to #717

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -2,6 +2,11 @@
 
 class User < ApplicationRecord
   attr_accessor :grant_permission_role
+  attr_accessor :pending_reviewer_invitations
+
+  before_create :set_pending_reviewer_invitations, prepend: true
+  after_commit  :process_pending_reviewer_invitations,  on: :create,
```
Comment (fanglinnw) [inline app/models/user.rb:8]: I think after_commit is not necessary in this case. after_commit is suitable for interacting with external system that can't be rolled back with in the database transaction. But it seems that processing pending invitation only involves database.

Comment (jseraf) [reply app/models/user.rb]: i think `after_create` failed in an earlier iteration but then `after_commit` didn't cause a problem moving forward.

**Comment by jseraf** [comment]
As it relates to a user, no. Grants would have pending/unconfirmed reviewer_invitations.

**Comment by fanglinnw** [comment]
> reviewer_invitations

so it seems unnecessary to have it as a virtual attribute, since it's not being and won't be used anywhere else other than the callbacks.

**Comment by jseraf** [comment]
@fanglinnw Something like this then?

```
after_create  :process_pending_reviewer_invitations
[...]
  def process_pending_reviewer_invitations
    #self.pending_reviewer_invitations.each do |invitation|
    GrantReviewer::Invitation.unconfirmed.where(email: email).each do |invitation|
      reviewer_assignment = GrantReviewer.new(grant: invitation.grant, reviewer: self)
      if reviewer_assignment.save
        invitation.update(confirmed_at: DateTime.now)
      end
    end
  end
```

**Comment by fanglinnw** [comment]
> @fanglinnw Something like this then?
> 
> ```
> after_create  :process_pending_reviewer_invitations
> [...]
>   def process_pending_reviewer_invitations
>     GrantReviewer::Invitation.unconfirmed.where(email: email).each do |invitation|
>       reviewer_assignment = GrantReviewer.new(grant: invitation.grant, reviewer: self)
>       if reviewer_assignment.save
>         invitation.update(confirmed_at: DateTime.now)
>       end
>     end
>   end
> ```

yes that's what I had in mind

**Review by mattbaumann1 — APPROVED** [review]

**Comment by jseraf** [comment]
most recent commit should say `PR feedback: change pending_reviewer_invitations to instance variable`

File: app/models/user.rb [hunk 2]
```diff
@@ -69,11 +69,11 @@ def get_role_by_grant(grant:)
   end
 
   def set_pending_reviewer_invitations
```
Comment (fanglinnw) [inline app/models/user.rb:71]: the method should be renamed to pending_Review_invitations

File: app/models/registered_user.rb [hunk 3]
```diff
@@ -4,7 +4,8 @@ class RegisteredUser < User
   devise :database_authenticatable, :confirmable, :registerable, :rememberable, :recoverable, :validatable
 
   after_initialize :set_uid, if: :new_record?
-  before_create    :confirm_invited_reviewers, unless: -> { self.pending_reviewer_invitations.empty? }
+  after_validation :confirm_invited_reviewers, on: :create,
```
Comment (fanglinnw) [inline app/models/registered_user.rb:7]: it should not directly use the instance variable here. Instead it should call the method which returns the varaible

**Review by fanglinnw — APPROVED** [review]

**Pull request merged**